src/Adapter/ContainerBuilder.php

```diff
@@ -38,35 +47,183 @@
+        //them at each container creation, this can't be compiled.
+        $this->loadDoctrineAnnotationMetadata();
+        $this->loadModulesAutoloader();
+
+        $container = $this->loadDumpedContainer();
+        if (null === $container) {
+            $container = $this->compileContainer();
+        }
+
+        return $container;
+    }
+
+    /**
+     * @return ContainerInterface|null
+     */
+    private function loadDumpedContainer()
+    {
+        $container = null;
+        if ($this->containerConfigCache->isFresh()) {
+            require_once $this->dumpFile;
+            $container = new $this->containerClassName();
         }
 
+        return $container;
+    }
+
+    /**
+     * @return SfContainerBuilder
+     *
+     * @throws \Exception
+     */
+    private function compileContainer()
+    {
         $container = new SfContainerBuilder();
+
         $container->addCompilerPass(new LegacyCompilerPass());
-        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
-        $env = $isDebug ? 'dev' : 'prod';
-        $servicesPath = _PS_CONFIG_DIR_ . "services/${name}/services_${env}.yml";
-        $loader->load($servicesPath);
+        $container->addCompilerPass(new LoadServicesFromModulesPass($this->containerName), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
+        $this->loadServices($container);
+
+        //Build extensions
+        $builderExtensions = [
+            new DoctrineBuilderExtension($this->environment),
+        ];
+        /** @var ContainerBuilderExtensionInterface $builderExtension */
+        foreach ($builderExtensions as $builderExtension) {
+            $builderExtension->build($container);
+        }
+
         $container->compile();
 
+        //Dump the container file
         $dumper = new PhpDumper($container);
-        file_put_contents($file, $dumper->dump(array('class' => $containerName)));
+        $this->containerConfigCache->write(
+            $dumper->dump(['class' => $this->containerClassName]),
+            $container->getResources()
+        );
 
         return $container;
     }
+
+    /**
+     * In symfony context doctrine classes (like Table, Entity, ...) are available thanks to
+     * the autoloader. In this specific context we don't have the general autoloader, so we need
+     * to include these classes manually. This is performed in Doctrine\ORM\Configuration::newDefaultAnnotationDriver
+     * which is called in Setup::createAnnotationMetadataConfiguration.
+     */
+    private function loadDoctrineAnnotationMetadata()
+    {
+        Setup::createAnnotationMetadataConfiguration([]);
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Exception
+     */
+    private function loadServices(SfContainerBuilder $container)
+    {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
+        $servicesPath = sprintf(
+            '%sservices/%s/services_%s.yml',
+                _PS_CONFIG_DIR_,
```

Comment:
wrong indent

Reply:
damn IDE.. (should I switch to emacs..?)

Reply:
You can try if you're a real man :trollface: